### PR TITLE
Make Steepest Descent a regular integrator

### PR DESCRIPTION
**Pull request by jngrad**
Description of changes:
- Remove the deprecated `minimize_energy` module
    - use `system.integrator.set_steepest_descent()` directly
- Simplify the MPI callback mechanism for Steepest Descent
    - broadcast struct parameters only once during initialization, instead of once per integration
    - call the `integrate()` MPI callback directly instead of delegating that call to the `steepest_descent()` MPI callback
- Document how to set up a custom convergence criterion

### CODE REVIEW

**Comment by jngrad** [comment]
This PR depends on #3880 and currently includes it. Here is the relevant [diff subset](https://github.com/jngrad/espresso/compare/fix-3850...jngrad:remove-minimize).
@christophlohrmann could you please have a look at the samples and tutorials?
```
samples/chamber_game.py
samples/drude_bmimpf6.py
samples/visualization_bonded.py
samples/visualization_cellsystem.py
samples/visualization_charged.py
samples/visualization_elc.py
samples/visualization_npt.py
doc/tutorials/02-charged_system/02-charged_system-1.ipynb
doc/tutorials/02-charged_system/02-charged_system-2.ipynb
doc/tutorials/02-charged_system/scripts/nacl.py
doc/tutorials/02-charged_system/scripts/nacl_units_confined.py
doc/tutorials/02-charged_system/scripts/nacl_units_confined_vis.py
doc/tutorials/04-lattice_boltzmann/04-lattice_boltzmann_part3.ipynb
doc/tutorials/04-lattice_boltzmann/scripts/04-lattice_boltzmann_part3_solution.py
```

**Comment by christophlohrmann** [comment]
From manual check: Visualizers look good, systems do not explode

File: doc/sphinx/running.rst [hunk 1]
```diff
@@ -308,11 +343,11 @@ The following minimal example illustrates how to use the SDM in |es|::
 
     import espressomd
     system = espressomd.System(box_l=[1.0, 1.0, 1.0])
+    system.periodicity = [False, False, False]
```
Comment (KaiSzuttor) [inline doc/sphinx/running.rst:347]: why?

File: src/core/integrators/steepest_descent.cpp [hunk 2]
```diff
@@ -100,15 +102,14 @@ bool steepest_descent_step(const ParticleRange &particles) {
...
   params.gamma = gamma;
-  params.max_steps = max_steps;
   params.max_displacement = max_displacement;
 }
 
-int steepest_descent(const int max_steps) {
-  params.max_steps = max_steps;
-  MPI_Bcast(&params, sizeof(SteepestDescentParameters), MPI_BYTE, 0, comm_cart);
-  return integrate(params.max_steps, -1);
+void mpi_bcast_steepest_descent_worker(int, int) {
```
Comment (KaiSzuttor) [inline src/core/integrators/steepest_descent.cpp:110]: please use the boost mpi wrapper

File: testsuite/python/integrator_exceptions.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+        self.system.integrator.set_brownian_dynamics()
+        with self.assertRaisesRegex(Exception, self.msg + 'The BD integrator requires the BD thermostat'):
+            self.system.integrator.run(0)
+
+    def test_brownian_integrator_brownian_thermostat(self):
+        self.system.thermostat.set_brownian(kT=1.0, gamma=1.0, seed=42)
+        self.system.integrator.set_brownian_dynamics()
+        self.assertCanIntegrate()
+
+    @utx.skipIfMissingFeatures("NPT")
```
Comment (KaiSzuttor) [inline testsuite/python/integrator_exceptions.py]: this test is unmaintainable

**Review by KaiSzuttor — CHANGES_REQUESTED** [review]
LGTM except for the long testcase that needs to be shortened somehow.

File: doc/sphinx/running.rst (hunk 1 above)
Comment (jngrad) [reply doc/sphinx/running.rst:347]: the SD method doesn't support periodic boundary conditions

File: src/core/integrators/steepest_descent.cpp (hunk 2 above)
Comment (jngrad) [reply src/core/integrators/steepest_descent.cpp:110]: done

File: testsuite/python/integrator_exceptions.py (hunk 3 above)
Comment (jngrad) [reply testsuite/python/integrator_exceptions.py]: simplified the test to do only one check per integrator, that should be sufficient

**Comment by jngrad** [comment]
There is a silent regression somewhere. I get random failures in the ROCm image for tests `dawaanr-and-dds-gpu` and `dawaanr-and-dds-gpu` that look like UB (logfiles [1](https://gitlab.icp.uni-stuttgart.de/espressomd/espresso/-/jobs/252133), [2](https://gitlab.icp.uni-stuttgart.de/espressomd/espresso/-/jobs/252141), [3](https://gitlab.icp.uni-stuttgart.de/espressomd/espresso/-/jobs/252142)), but I can't reproduce them on `lama` nor in the ROCm docker image on `lama`. They are only reproducible in the docker image on `coyote10`, but not every time, and in GDB the backtrace is sometimes squashed. The few times I got a usable backtrace in GDB, it was about incorrect particle ids and double free of the cellsystem particle list (`dawaanr-and-dds-gpu` traces [1.log](https://github.com/espressomd/espresso/files/5226097/dawaanr-and-dds-gpu-1.log), [2.log](https://github.com/espressomd/espresso/files/5226098/dawaanr-and-dds-gpu-2.log)).

The regression is independent of 022ee4e91 and independent of another silent regression that was fixed by 112bd65c.

**Comment by jngrad** [comment]
Even more failures after the bugfix in 112bd65c824 ([logfile](https://gitlab.icp.uni-stuttgart.de/espressomd/espresso/-/jobs/252243)):
```
	 32 - dawaanr-and-dds-gpu (Failed)
	 33 - dawaanr-and-bh-gpu (Failed)
	 34 - dds-and-bh-gpu (Failed)
```

This is not a regression, the bug is already present on the `python` branch. Tested on fe967ea4ae with the ROCm image on coyote11: `dawaanr-and-dds-gpu`, `dawaanr-and-bh-gp` and `dds-and-bh-gpu` all fail randomly. It's also failing in other PRs.

File: doc/sphinx/running.rst [hunk 4]
```diff
@@ -259,20 +261,53 @@ Usage example::
...
-available to set up the steepest descent integrator while preserving the
-original integrator. The snippet above can be rewritten to::
+Using a custom convergence criterion
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The ``f_max`` parameter can be set to zero to prevent the integrator from
+halting when a specific force/torque is reached. The integration can then
+be carried out in a loop with a custom convergence criterion::
+
+    max_sigma = 5
```
Comment (KaiSzuttor) [inline doc/sphinx/running.rst]: wouldn't you want to take the smallest LJ sigma here?

File: src/core/integrators/steepest_descent.cpp [hunk 5]
```diff
@@ -100,15 +101,13 @@ bool steepest_descent_step(const ParticleRange &particles) {
...
+  namespace mpi = boost::mpi;
+  mpi::broadcast(comm_cart, params, 0);
```
Comment (KaiSzuttor) [inline src/core/integrators/steepest_descent.cpp]: the alias is probably not needed, since you even add more text :D. Why is the return gone?

**Review by KaiSzuttor — COMMENTED** [review]
after my comments have been addressed, this can be merged

File: src/core/integrators/steepest_descent.cpp (hunk 5 above)
Comment (jngrad) [reply src/core/integrators/steepest_descent.cpp]: the deleted return belonged to a different function that has been removed

File: doc/sphinx/running.rst (hunk 4 above)
Comment (jngrad) [reply doc/sphinx/running.rst]: * distance criterion was clarified
* offline discussion: energy convergence vs. minimal distance, would require touching a lot of samples and tutorials, some of which are candidate for removal once tutorial 02 part 1 and its solution scripts get replaced by the charged rod tutorial

**Review by KaiSzuttor — APPROVED** [review]

**Pull request merged**